From c404b2e11d49a660c29a7b63b1843baa15be8af6 Mon Sep 17 00:00:00 2001
From: Sagar Shelke <shelkesagar29@yahoo.com>
Date: Tue, 1 Jul 2025 21:53:46 +0000
Subject: [PATCH] Apply patch

---
 .../optimization/StablehloAggressiveSimplificationPatterns.td  | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/stablehlo/transforms/optimization/StablehloAggressiveSimplificationPatterns.td b/stablehlo/transforms/optimization/StablehloAggressiveSimplificationPatterns.td
index efae71b2..992ac5de 100644
--- a/stablehlo/transforms/optimization/StablehloAggressiveSimplificationPatterns.td
+++ b/stablehlo/transforms/optimization/StablehloAggressiveSimplificationPatterns.td
@@ -98,7 +98,8 @@ def SortedDims : AttrConstraint<
     "is sorted dimensions">;
 
 def ZeroExtent : AttrConstraint<
-    CPred<"cast<DenseIntElementsAttr>($_self).getNumElements() == 0">,
+    CPred<"cast<ElementsAttr>($_self).getNumElements() == 0 &&"
+          "isa<IntegerType>(cast<ElementsAttr>($_self).getElementType())">,
     "is zero extent">;
 
 ///////////
-- 
2.48.1

