From ac6ac8785d9a6429304d1ccd6f9c894373182dc0 Mon Sep 17 00:00:00 2001
From: Sagar Shelke <shelkesagar29@yahoo.com>
Date: Tue, 1 Jul 2025 21:54:20 +0000
Subject: [PATCH] Apply patch

---
 stablehlo/transforms/conversions/CMakeLists.txt | 1 -
 1 file changed, 1 deletion(-)

diff --git a/stablehlo/transforms/conversions/CMakeLists.txt b/stablehlo/transforms/conversions/CMakeLists.txt
index e1da2c8b..7aac80f1 100644
--- a/stablehlo/transforms/conversions/CMakeLists.txt
+++ b/stablehlo/transforms/conversions/CMakeLists.txt
@@ -18,7 +18,6 @@ add_mlir_library(StablehloTypeConversion
   TypeConversion.cpp
 
   LINK_LIBS PUBLIC
-  LLVMSupport
   MLIRIR
   MLIRSupport
   MLIRTransformUtils
-- 
2.48.1

