From d679ee8c7978fe63321e90c5c6583b604ca3d1a5 Mon Sep 17 00:00:00 2001
From: Sagar Shelke <shelkesagar29@yahoo.com>
Date: Tue, 1 Jul 2025 00:25:43 +0000
Subject: [PATCH] Apply patch 0009

---
 mlir/lib/Target/LLVMIR/DebugTranslation.cpp | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/mlir/lib/Target/LLVMIR/DebugTranslation.cpp b/mlir/lib/Target/LLVMIR/DebugTranslation.cpp
index 1d3ed6f3262f..93e1d08faf4f 100644
--- a/mlir/lib/Target/LLVMIR/DebugTranslation.cpp
+++ b/mlir/lib/Target/LLVMIR/DebugTranslation.cpp
@@ -547,6 +547,14 @@ llvm::DILocation *DebugTranslation::translateLoc(Location loc,
         llvm::DILocation::get(llvmCtx, fileLoc.getLine(), fileLoc.getColumn(),
                               scope, const_cast<llvm::DILocation *>(inlinedAt));
 
+  } else if (auto fileLoc = dyn_cast<FileLineColRange>(loc)) {
+    // A scope of a DILocation cannot be null.
+    if (!scope)
+      return nullptr;
+    llvmLoc = llvm::DILocation::get(llvmCtx, fileLoc.getStartLine(),
+                                    fileLoc.getStartColumn(), scope,
+                                    const_cast<llvm::DILocation *>(inlinedAt));
+
   } else if (auto fusedLoc = dyn_cast<FusedLoc>(loc)) {
     ArrayRef<Location> locations = fusedLoc.getLocations();
 
-- 
2.48.1

