From bcf9dd3472cb5b45a25843f3956fb92a2b38e9b3 Mon Sep 17 00:00:00 2001
From: Christopher Bate <cbate@nvidia.com>
Date: Fri, 21 Mar 2025 16:40:57 +0000
Subject: [PATCH 2/3] Make compatible with more recent Stablehlo version

---
 lib/InitAll.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/lib/InitAll.cpp b/lib/InitAll.cpp
index d9096929..89773e60 100644
--- a/lib/InitAll.cpp
+++ b/lib/InitAll.cpp
@@ -20,6 +20,7 @@
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
 #include "mlir/Dialect/Tensor/IR/TensorInferTypeOpInterfaceImpl.h"
 #include "mlir/IR/Dialect.h"
+#include "stablehlo/transforms/optimization/Passes.h"
 #include "torch-mlir-dialects/Dialect/TMTensor/IR/TMTensorDialect.h"
 #include "torch-mlir-dialects/Dialect/TMTensor/Transforms/Passes.h"
 #include "torch-mlir/Conversion/Passes.h"
@@ -32,6 +33,7 @@
 
 #ifdef TORCH_MLIR_ENABLE_STABLEHLO
 #include "stablehlo/conversions/linalg/transforms/Passes.h"
+#include "stablehlo/transforms/optimization/Passes.h"
 #include "stablehlo/transforms/Passes.h"
 #endif
 
@@ -72,7 +74,7 @@ void mlir::torch::registerAllPasses() {
 
 #ifdef TORCH_MLIR_ENABLE_STABLEHLO
   mlir::stablehlo::registerStablehloLegalizeToLinalgPass();
-  mlir::stablehlo::registerStablehloAggressiveSimplificationPass();
+  mlir::stablehlo::registerOptimizationPasses();
   mlir::stablehlo::registerStablehloRefineShapesPass();
   mlir::stablehlo::registerStablehloConvertToSignlessPass();
   mlir::stablehlo::registerShapeLegalizeToStablehloPass();
-- 
2.46.0

