diff --git a/stablehlo/integrations/c/CMakeLists.txt b/stablehlo/integrations/c/CMakeLists.txt
index 74595438..400a1121 100644
--- a/stablehlo/integrations/c/CMakeLists.txt
+++ b/stablehlo/integrations/c/CMakeLists.txt
@@ -28,7 +28,6 @@ add_mlir_public_c_api_library(ChloCAPI
 
   LINK_LIBS PUBLIC
   ChloOps
-  LLVMSupport
 )
 
 add_mlir_public_c_api_library(StablehloCAPI
@@ -40,10 +39,9 @@ add_mlir_public_c_api_library(StablehloCAPI
   StablehloTypes.cpp
 
   LINK_LIBS PUBLIC
-  LLVMSupport
   MLIRCAPIIR
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
   StablehloOps
   StablehloPasses
   StablehloPortableApi
diff --git a/stablehlo/reference/CMakeLists.txt b/stablehlo/reference/CMakeLists.txt
index b3a406b0..b6dcac90 100644
--- a/stablehlo/reference/CMakeLists.txt
+++ b/stablehlo/reference/CMakeLists.txt
@@ -20,7 +20,7 @@ add_mlir_library(StablehloReferenceApi
   MLIRIR
   MLIRParser
   MLIRPass
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
   InterpreterOps
   StablehloPasses
   StablehloReferenceConfiguration
@@ -40,7 +40,7 @@ add_mlir_library(StablehloReferenceAxes
 
   LINK_LIBS PUBLIC
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
 )
 
 add_mlir_library(StablehloReferenceConfiguration
@@ -48,7 +48,7 @@ add_mlir_library(StablehloReferenceConfiguration
   Configuration.cpp
 
   LINK_LIBS PUBLIC
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
   StablehloReferenceErrors
   StablehloReferenceProcess
   StablehloReferenceScope
@@ -60,7 +60,7 @@ add_mlir_library(StablehloReferenceElement
 
   LINK_LIBS PUBLIC
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
   StablehloReferenceTypes
 )
 
@@ -70,7 +70,7 @@ add_mlir_library(StablehloReferenceErrors
 
   LINK_LIBS PUBLIC
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
 )
 
 add_mlir_library(StablehloReferenceIndex
@@ -79,7 +79,7 @@ add_mlir_library(StablehloReferenceIndex
 
   LINK_LIBS PUBLIC
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
 )
 
 set(LLVM_TARGET_DEFINITIONS InterpreterOps.td)
@@ -102,7 +102,7 @@ add_mlir_dialect_library(InterpreterOps
   StablehloReferenceOps
   StablehloReferenceProcessGrid
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
 )
 
 set(LLVM_TARGET_DEFINITIONS InterpreterPasses.td)
@@ -120,7 +120,7 @@ add_mlir_library(InterpreterPasses
   InterpreterOps
   MLIRIR
   MLIRPass
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
   MLIRTransforms
   MLIRTransformUtils
 )
@@ -169,9 +169,8 @@ add_mlir_library(StablehloReferenceProcessGrid
   ProcessGrid.cpp
 
   LINK_LIBS PUBLIC
-  LLVMSupport
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
   StablehloReferenceTensor
 )
 
@@ -183,7 +182,7 @@ add_mlir_library(StablehloReferenceScope
   StablehloReferenceValue
   StablehloReferenceTensor
   StablehloReferenceToken
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
 )
 
 add_mlir_library(StablehloReferenceTensor
@@ -205,7 +204,7 @@ add_mlir_library(StablehloReferenceToken
 
   LINK_LIBS PUBLIC
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
   StablehloOps
 )
 
@@ -215,7 +214,7 @@ add_mlir_library(StablehloReferenceTypes
 
   LINK_LIBS PUBLIC
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
 )
 
 add_mlir_library(StablehloReferenceValue
@@ -224,7 +223,7 @@ add_mlir_library(StablehloReferenceValue
 
   LINK_LIBS PUBLIC
   MLIRIR
-  MLIRSupport
+  MLIRSupport  # LLVMSupport is automatically added in AddLLVM.cmake.
   StablehloReferenceTensor
   StablehloReferenceToken
 )
