diff --git a/stablehlo/transforms/StablehloAggressiveSimplification.cpp b/stablehlo/transforms/StablehloAggressiveSimplification.cpp
index 8e9332f3..2bfdf34f 100644
--- a/stablehlo/transforms/StablehloAggressiveSimplification.cpp
+++ b/stablehlo/transforms/StablehloAggressiveSimplification.cpp
@@ -39,6 +39,40 @@ namespace stablehlo {
 #include "stablehlo/transforms/Passes.h.inc"
 
 namespace {
+
+/// Does the same as PatternRewriter::replaceOpWithNewOp, but with a twist.
+///
+/// Sometimes, we want to replace an op with a new op and simultaneously refine
+/// the result type from a dynamically-shaped type to a statically-shaped type.
+/// (Search for usages of this function for examples).
+//
+/// Oftentimes, this works just fine because HLO is designed to accommodate
+/// this kind of type refinements. But sometimes, this doesn't work - when
+/// the op is used outside of the HLO dialect (e.g. in func.return). In these
+/// cases, we insert a tensor.cast to smooth things out.
+template <typename OpTy, typename... Args>
+static OpTy refineOpWithNewOp(PatternRewriter &rewriter, Operation *op,
+                              Args &&...args) {
+  auto newOp = rewriter.create<OpTy>(op->getLoc(), std::forward<Args>(args)...);
+
+  llvm::SmallVector<Value> replacementResults;
+  assert(op->getNumResults() == newOp->getNumResults() &&
+         "replacement op doesn't match results of original op");
+  for (auto [opResult, newOpResult] :
+       llvm::zip(op->getResults(), newOp->getResults())) {
+    Value replacementResult = newOpResult;
+    if (llvm::any_of(opResult.getUsers(), [&](Operation *user) {
+          return user->getDialect() != op->getDialect();
+        }))
+      replacementResult = rewriter.create<mlir::tensor::CastOp>(
+          op->getLoc(), opResult.getType(), newOpResult);
+    replacementResults.push_back(replacementResult);
+  }
+
+  rewriter.replaceOp(op, replacementResults);
+  return newOp;
+}
+
 // This is an upper limit on how many elements canonicalization patterns are
 // allowed to materialize as new constants.
 constexpr int64_t kFoldOpEltLimit = 65536;
@@ -119,7 +153,7 @@ struct AddOpCanon final : OpRewritePattern<mlir::stablehlo::AddOp> {
     if (TypedAttr res;
         lhsAttr && rhsAttr &&
         (res = foldBinaryOpIntOrFloat(lhsAttr, rhsAttr, std::plus<>{}))) {
-      rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(op, res);
+      refineOpWithNewOp<mlir::stablehlo::ConstantOp>(rewriter, op, res);
       return success();
     }
 
@@ -157,7 +191,7 @@ struct SubtractOpCanon final : OpRewritePattern<mlir::stablehlo::SubtractOp> {
     if (TypedAttr res;
         lhsAttr && rhsAttr &&
         (res = foldBinaryOpIntOrFloat(lhsAttr, rhsAttr, std::minus<>{}))) {
-      rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(op, res);
+      refineOpWithNewOp<mlir::stablehlo::ConstantOp>(rewriter, op, res);
       return success();
     }
 
@@ -208,7 +242,7 @@ struct MulOpCanon final : OpRewritePattern<mlir::stablehlo::MulOp> {
     if (TypedAttr res;
         lhsAttr && rhsAttr &&
         (res = foldBinaryOpIntOrFloat(lhsAttr, rhsAttr, std::multiplies<>{}))) {
-      rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(op, res);
+      refineOpWithNewOp<mlir::stablehlo::ConstantOp>(rewriter, op, res);
       return success();
     }
 
@@ -299,15 +333,16 @@ struct CompareOpCanon final : OpRewritePattern<mlir::stablehlo::CompareOp> {
         case ComparisonDirection::EQ:
         case ComparisonDirection::GE:
         case ComparisonDirection::LE: {
-          rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(
-              op, SplatElementsAttr::get(type, rewriter.getBoolAttr(true)));
+          refineOpWithNewOp<mlir::stablehlo::ConstantOp>(
+              rewriter, op,
+              SplatElementsAttr::get(type, rewriter.getBoolAttr(true)));
           return success();
         }
         case ComparisonDirection::GT:
         case ComparisonDirection::LT:
         case ComparisonDirection::NE: {
-          rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(
-              op, rewriter.getZeroAttr(type));
+          refineOpWithNewOp<mlir::stablehlo::ConstantOp>(
+              rewriter, op, rewriter.getZeroAttr(type));
           return success();
         }
       }
@@ -336,7 +371,7 @@ struct CompareOpCanon final : OpRewritePattern<mlir::stablehlo::CompareOp> {
              [direction, kind = *compType](const APInt &a, const APInt &b) {
                return calculateComp(kind, direction, a, b);
              }))) {
-      rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(op, res);
+      refineOpWithNewOp<mlir::stablehlo::ConstantOp>(rewriter, op, res);
       return success();
     }
 
@@ -349,8 +384,6 @@ struct SelectOpCanon final : OpRewritePattern<mlir::stablehlo::SelectOp> {
 
   LogicalResult matchAndRewrite(mlir::stablehlo::SelectOp op,
                                 PatternRewriter &rewriter) const override {
-    RankedTensorType type = op.getType();
-
     Value trueVal = op.getOnTrue();
     Value falseVal = op.getOnFalse();
 
@@ -373,7 +406,9 @@ struct SelectOpCanon final : OpRewritePattern<mlir::stablehlo::SelectOp> {
 
     // Handle elementwise selection when both outcomes are also constants. This
     // will create a new, likely non-splat constant.
-    if (cond.getNumElements() > kFoldOpEltLimit) return failure();
+    if (!cond.getShapedType().hasStaticShape() ||
+        cond.getNumElements() > kFoldOpEltLimit)
+      return failure();
 
     ElementsAttr trueAttr;
     if (!matchPattern(trueVal, m_Constant(&trueAttr))) return failure();
@@ -389,8 +424,10 @@ struct SelectOpCanon final : OpRewritePattern<mlir::stablehlo::SelectOp> {
       newValues.push_back(condElem ? trueElem : falseElem);
     }
 
-    rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(
-        op, DenseElementsAttr::get(type, newValues));
+    refineOpWithNewOp<mlir::stablehlo::ConstantOp>(
+        rewriter, op,
+        DenseElementsAttr::get(cast<RankedTensorType>(trueVal.getType()),
+                               newValues));
     return success();
   }
 };
@@ -465,9 +502,10 @@ struct BroadcastInDimOpCanon final
     // Handle splat broadcasts.
     if (SplatElementsAttr cstAttr;
         matchPattern(operand, m_Constant(&cstAttr))) {
-      rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(
-          op, SplatElementsAttr::get(op.getType(),
-                                     cstAttr.getSplatValue<Attribute>()));
+      refineOpWithNewOp<mlir::stablehlo::ConstantOp>(
+          rewriter, op,
+          SplatElementsAttr::get(op.getType(),
+                                 cstAttr.getSplatValue<Attribute>()));
       return success();
     }
 
@@ -538,8 +576,8 @@ struct ConcatenateOpCanon final
     }
 
     assert(newElems.size() == numElems);
-    rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(
-        op, DenseElementsAttr::get(op.getType(), newElems));
+    refineOpWithNewOp<mlir::stablehlo::ConstantOp>(
+        rewriter, op, DenseElementsAttr::get(op.getType(), newElems));
     return success();
   }
 };
@@ -557,39 +595,6 @@ struct ConvertOpCanon final : OpRewritePattern<mlir::stablehlo::ConvertOp> {
   }
 };
 
-/// Does the same as PatternRewriter::replaceOpWithNewOp, but with a twist.
-///
-/// Sometimes, we want to replace an op with a new op and simultaneously refine
-/// the result type from a dynamically-shaped type to a statically-shaped type.
-/// (Search for usages of this function for examples).
-//
-/// Oftentimes, this works just fine because HLO is designed to accommodate
-/// this kind of type refinements. But sometimes, this doesn't work - when
-/// the op is used outside of the HLO dialect (e.g. in func.return). In these
-/// cases, we insert a tensor.cast to smooth things out.
-template <typename OpTy, typename... Args>
-static OpTy refineOpWithNewOp(PatternRewriter &rewriter, Operation *op,
-                              Args &&...args) {
-  auto newOp = rewriter.create<OpTy>(op->getLoc(), std::forward<Args>(args)...);
-
-  llvm::SmallVector<Value> replacementResults;
-  assert(op->getNumResults() == newOp->getNumResults() &&
-         "replacement op doesn't match results of original op");
-  for (auto [opResult, newOpResult] :
-       llvm::zip(op->getResults(), newOp->getResults())) {
-    Value replacementResult = newOpResult;
-    if (llvm::any_of(opResult.getUsers(), [&](Operation *user) {
-          return user->getDialect() != op->getDialect();
-        }))
-      replacementResult = rewriter.create<mlir::tensor::CastOp>(
-          op->getLoc(), opResult.getType(), newOpResult);
-    replacementResults.push_back(replacementResult);
-  }
-
-  rewriter.replaceOp(op, replacementResults);
-  return newOp;
-}
-
 /// If a DynamicBroadCastInDimOp is not actually dynamic, use an ordinary
 /// BroadcastInDimOp.
 struct DynamicBroadcastInDimOpNotActuallyDynamic final
@@ -924,8 +929,8 @@ struct GetDimensionSizeOpCanon final
 
     auto elemTy = cast<IntegerType>(op.getType().getElementType());
     IntegerAttr elemVal = rewriter.getIntegerAttr(elemTy, dimSize);
-    rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(
-        op, DenseElementsAttr::get(op.getType(), elemVal));
+    refineOpWithNewOp<mlir::stablehlo::ConstantOp>(
+        rewriter, op, DenseElementsAttr::get(op.getType(), elemVal));
     return success();
   }
 };
@@ -1017,16 +1022,17 @@ struct ReshapeOpCanon final : OpRewritePattern<mlir::stablehlo::ReshapeOp> {
     if (!matchPattern(op.getOperand(), m_Constant(&cstAttr))) return failure();
 
     if (auto splat = dyn_cast<SplatElementsAttr>(cstAttr)) {
-      rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(
-          op, SplatElementsAttr::get(op.getType(),
-                                     splat.getSplatValue<Attribute>()));
+      refineOpWithNewOp<mlir::stablehlo::ConstantOp>(
+          rewriter, op,
+          SplatElementsAttr::get(op.getType(),
+                                 splat.getSplatValue<Attribute>()));
       return success();
     }
 
     auto elements =
         llvm::to_vector_of<Attribute>(cstAttr.getValues<Attribute>());
-    rewriter.replaceOpWithNewOp<mlir::stablehlo::ConstantOp>(
-        op, DenseElementsAttr::get(op.getType(), elements));
+    refineOpWithNewOp<mlir::stablehlo::ConstantOp>(
+        rewriter, op, DenseElementsAttr::get(op.getType(), elements));
     return success();
   }
 };
